chaincode/src/swaps/fillTokenSwap.ts

```diff
@@ -0,0 +1,393 @@
+
+  // If we have any errors, don't proceed to check the spendable quantity and
+  // issue the transaction
+  if (chainValidationErrors.length === 0) {
+    // Check from balances
+    for (let index = 0; index < wantedTokenInstances.length; index++) {
+      const toToken = wantedTokenInstances[index];
+      const balance = await fetchOrCreateBalance(ctx, filledBy, toToken);
+
+      if (toToken.isNonFungible) {
+        if (!balance.isInstanceSpendable(toToken.instance, ctx.txUnixTime)) {
+          chainValidationErrors.push(
+            `${filledBy} has insufficient balance to spend 1 of token ${toToken.GetCompositeKeyString()}. SwapRequestId ${swapRequestId}`
+          );
+        }
+      } else {
+        const currentSpendableQuantity = balance.getQuantityTotal();
+        const currentQuantity = tokenSwap.wanted[index].quantity.times(uses);
+
+        if (!currentSpendableQuantity.toNumber() || currentSpendableQuantity.isLessThan(currentQuantity)) {
+          chainValidationErrors.push(
+            `${filledBy} has insufficient balance to spend ${currentQuantity} of token ${toToken.GetCompositeKeyString()}. SwapRequestId ${swapRequestId}`
+          );
+        }
+      }
+    }
+  }
+
+  // If we have any errors after checking balances, don't proceed
+  if (chainValidationErrors.length === 0) {
+    // Transfer From Tokens
+    for (let index = 0; index < tokenSwap.offered.length; index++) {
+      const fromTokenId = tokenSwap.offered[index].tokenInstance;
+      const currentQuantity = tokenSwap.offered[index].quantity.times(uses);
+
+      ///////////////////////////////
+      // We are skipping user validation because the quantities have already
+      // been validated by the swap
+      ///////////////////////////////
+      // TODO: This should eventually use the transfer Token function when there is a better solution for normalizing
+      // validation for swaps and transfers
+      await swapToken(ctx, tokenSwap.offeredBy, filledBy, fromTokenId, currentQuantity, swapRequestId).catch(
+        (e) => {
+          const chainError = ChainError.from(e);
+          throw new SwapTokenFailedError(
+            chainError.message + `SwapRequestId ${swapRequestId}`,
+            chainError.payload
+          );
+        }
+      );
+    }
+
+    // Transfer To Tokens - note that filledBy user doesn't have locked tokens
+    for (let index = 0; index < tokenSwap.wanted.length; index++) {
+      const toTokenId = tokenSwap.wanted[index].tokenInstance;
+      const currentQuantity = tokenSwap.wanted[index].quantity.times(uses);
+
+      if (filledOnBehalfOfOwner) {
+        // todo: unsure why verifyAndUseAllowances() requires *both* the tokenInstanceKey and tokenInstance be supplied -
+        // its code appears to just need the string properties not the actual TokenInstance off chain
+        // that might be worth reworking in the future
```

Comment:
could you gather all of those problematic TODOs, and create a new issue for them? I believe it's a good way of enforcing more thorough review (by doing) of the business logic behind swaps

Reply:
Agree, I have written up a new issue here:

https://github.com/GalaChain/sdk/issues/470

It summarizes both high level goals of a review/refactor given new features introduced in the SDK since it was originally written, and it summarizes your comments on this pull request into a bulleted list.